Set default status in waitpid when pid = 0.

toots opened the pull request:
Refined PR based on comments at #2222

toots committed:
Set default status in waitpid when pid = 0.

stedolan commented:
This looks good to me. @dra27 ?

xavierleroy commented on testsuite/tests/lib-unix/common/wait_nohang.ml:
```diff
@@ -15,3 +15,19 @@ let () =
+  let pid, status = Unix.waitpid [WNOHANG] pid in
+  assert (pid = 0);
+  assert (status = WEXITED 0);
+  while fst (Unix.waitpid [WNOHANG] pid) = 0 do
```
This test looks very fragile to me.  On a busy machine, the `sleep 0.01` could finish before the tester (the present OCaml code) gets to run the `Unix.waitpid`.  Could you please write a test that has no race conditions?  E.g.
* Create a pipe
* Run `cat` with stdin = the pipe and stdout = /dev/null
* Check that `Unix.waitpid [WNOHANG] pid` is pid = 0, status = WEXITED 0.
* Close the pipe, causing `cat` to terminate.
* Check again  `Unix.waitpid [WNOHANG] pid`  in a loop like in the current code.

xavierleroy reviewed (commented):
The code is obviously good.  The test worries me, see comment below.

xavierleroy committed:
Improved wait_nohang test

xavierleroy commented:
I pushed on your branch  what I think is a better test, and will now give it a round of precheck CI.

xavierleroy commented:
The Windows ports are unhappy.  Investigation in progress.

xavierleroy committed:
Improved wait_nohang test, continued

Under Windows, there is a mixture of \n and \r\n end of lines that
cause the test to fail.  Simplify the test to avoid this.

xavierleroy committed:
Improved wait_nohang test, again and again

Invoke the reflector through Unix.open_process_out, this seems to work
better under Windows.

xavierleroy committed:
Merge branch 'trunk' into waitpid-nohang

xavierleroy commented:
After much CI and tweaking of the test, I declare success and will merge this PR.

toots commented:
Great, thanks for this!
